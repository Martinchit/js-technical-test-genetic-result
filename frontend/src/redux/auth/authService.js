import axios from 'axios';

export const logInService = async (email, password) => {
  const reqBody = {
    email,
    password,
  };
  const { data } = await axios({
    method: 'POST',
    baseURL: `${process.env.REACT_APP_API_ENDPOINT}`,
    url: '/api/auth/log-in',
    data: {
      ...reqBody,
    },
  });
  return data.token;
};
