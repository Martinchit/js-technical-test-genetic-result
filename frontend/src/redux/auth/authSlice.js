import { createSlice } from '@reduxjs/toolkit';
import { logInService } from './authService';
import { updateObject } from '../../core/lib/utils/updateObject';
import {
  setAuthToken,
  removeAuthToken,
  retrieveAuthToken,
} from '../../core/lib/utils/localStorage';

export const initialState = {
  isLocalStorageChecked: false,
  logInProcessing: false,
  loggedIn: false,
  logInError: null,
  token: null,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    reset: {
      reducer: (state) => {
        return updateObject(state, { logInError: null, logInProcessing: false });
      },
    },
    logInProcessing: {
      reducer: (state) => {
        return updateObject(state, { logInProcessing: true });
      },
    },
    logInSuccess: {
      reducer: (state, action) => {
        const { token } = action['payload'];
        return updateObject(state, {
          token,
          loggedIn: true,
          logInProcessing: false,
        });
      },
      prepare: (token) => ({
        payload: {
          token,
        },
      }),
    },
    logInError: {
      reducer: (state, action) => {
        const { error } = action['payload'];
        return updateObject(state, { logInError: error, logInProcessing: false });
      },
      prepare: (error) => ({ payload: { error } }),
    },
    logOut: (state) => {
      removeAuthToken();
      return updateObject(state, {
        token: null,
        loggedIn: false,
      });
    },
    localStorageCheckCompleted: {
      reducer: (state, action) => {
        const { token } = action['payload'];
        return updateObject(state, {
          token,
          isLocalStorageChecked: true,
          loggedIn: token ? true : false,
        });
      },
      prepare: (token) => ({ payload: { token } }),
    },
  },
});

export const {
  isLocalStorageChecked,
  logInSuccess,
  logInError,
  loggedIn,
  logOut,
  reset,
  logInProcessing,
  localStorageCheckCompleted,
} = authSlice.actions;

export default authSlice.reducer;

export const logInAction = (email, password) => async (dispatch) => {
  dispatch(reset());
  dispatch(logInProcessing());
  try {
    const token = await logInService(email, password);
    setAuthToken(token);
    dispatch(logInSuccess(token));
  } catch (err) {
    const { data } = err.response;
    dispatch(logInError(data.description));
  }
};

export const checkLocalStorageAction = () => (dispatch) => {
  const token = retrieveAuthToken();
  dispatch(localStorageCheckCompleted(token));
};
