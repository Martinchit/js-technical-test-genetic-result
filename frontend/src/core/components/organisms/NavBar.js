import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { Dropdown, Icon, Menu } from 'semantic-ui-react';
import { ImCool } from 'react-icons/im';
import { logOut } from '../../../redux/auth/authSlice';

export const NavBar = ({ loggedIn }) => {
  const dispatch = useDispatch();

  const logOutAction = React.useCallback(() => {
    dispatch(logOut());
  }, [dispatch]);

  const navButtons = React.useMemo(
    () =>
      loggedIn
        ? [
            {
              text: 'Result',
              name: 'result',
              url: '/result',
            },
          ]
        : [],
    [loggedIn]
  );

  return (
    <Menu stackable size="massive" fixed="top" id="navbar">
      <Menu.Item as={Link} to="/">
        <ImCool size="1.5em" />
      </Menu.Item>
      {loggedIn ? (
        <React.Fragment>
          {navButtons.map(({ text, name, url }) => (
            <Menu.Item key={name} name={name} as={Link} to={url}>
              {text}
            </Menu.Item>
          ))}
          <Menu.Menu position="right">
            <Menu.Item name="name" as={Dropdown} trigger={<Icon name="user" />}>
              <Dropdown.Menu>
                <Dropdown.Item text="Profile" as={Link} to="/profile" />
                <Dropdown.Item
                  text="Log Out"
                  onClick={logOutAction}
                  id="navbar-logout-button"
                />
              </Dropdown.Menu>
            </Menu.Item>
          </Menu.Menu>
        </React.Fragment>
      ) : null}
    </Menu>
  );
};

NavBar.propTypes = {
  loggedIn: PropTypes.bool,
};

NavBar.defaultProps = {
  loggedIn: false,
};
