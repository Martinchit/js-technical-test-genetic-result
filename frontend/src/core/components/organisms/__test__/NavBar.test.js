import React from "react";
import { shallow, mount } from 'enzyme';
import * as redux from 'react-redux';
import { Dropdown, Icon, Menu } from 'semantic-ui-react';
import 'jest-styled-components';

import { NavBar } from "../NavBar";
import { store } from '../../../../redux/store';
import { logInSuccess } from '../../../../redux/auth/authSlice';

describe("NavBar", () => {
  let spyOnUseDispatch;
  let mockDispatch;

  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders correctly", () => {
    const wrapper = shallow(<NavBar />)
    expect(wrapper).toMatchSnapshot()
  })

  it("renders correctly after log in", () => {
    const wrapper = shallow(<NavBar loggedIn />)
    expect(wrapper.find(Menu.Item).length).toBe(3)
    expect(wrapper.find(Dropdown.Menu).length).toBe(1)
    expect(wrapper.find(Dropdown.Item).length).toBe(2)
    wrapper.find('#navbar-logout-button').simulate('click');
    expect(mockDispatch.mock.calls[0][0]).toEqual({ type: 'auth/logOut', payload: undefined })
  })
})