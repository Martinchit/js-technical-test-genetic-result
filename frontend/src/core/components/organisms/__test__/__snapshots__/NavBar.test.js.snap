// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders correctly 1`] = `ShallowWrapper {}`;

exports[`NavBar renders correctly after log in 1`] = `ShallowWrapper {}`;
